rekall/plugins/addrspaces/pmem.py

```diff
@@ -0,0 +1,341 @@
+PMEM_IOCTL_GET_MMAP_DESC_SIZE = _IOR(PMEM_IOCTL_BASE,
+                                     PMEM_GET_MMAP_DESC_SIZE,
+                                     PMEM_MMAP_DESC_SIZE_TYPE)
+
+
+# Fills the buffer with the address of the DTB.
+PMEM_IOCTL_GET_DTB = _IOR(PMEM_IOCTL_BASE,
+                          PMEM_GET_DTB,
+                          PMEM_DTB_TYPE)
+
+
+# Changes the method used by the Pmem driver. (PTE is the default.)
+PMEM_IOCTL_SET_MMAP_METHOD = _IOW(PMEM_IOCTL_BASE,
+                                  PMEM_SET_MMAP_METHOD,
+                                  PMEM_MMAP_METHOD_TYPE)
+
+
+def pmem_get_mmap_size(fd):
+    """Ask the Pmem driver for mmap size (number of entries)."""
+    buf = array.array("I", [0])
+    err = fcntl.ioctl(fd, PMEM_IOCTL_GET_MMAP_SIZE, buf, True)
+    if err:
+        raise IOError("Error (%d) getting mmap size." % err)
+
+    return buf[0]
+
+
+def pmem_get_mmap_desc_size(fd):
+    """Ask the Pmem driver for size of an efi descriptor.
+
+    Note that this must absolutely ALWAYS return 0x30, otherwise we're
+    dealing with some future version fo EFI we know nothing about.
+    """
+    buf = array.array("I", [0])
+    err = fcntl.ioctl(fd, PMEM_IOCTL_GET_MMAP_DESC_SIZE, buf, True)
+    if err:
+        raise IOError("Error (%d) getting mmap desc size." % err)
+
+    return buf[0]
+
+
+def pmem_get_mmap(fd):
+    """Ask the Pmem driver for the physical address map.
+
+    Returns:
+    ========
+
+    Tuple of (mmap, no_entries, entry_size).
+
+    Each multiple of entry_size in the mmap_buffer is a valid
+    EFI_MEMORY_DESCRIPTOR (defined below).
+
+    Raises:
+    =======
+
+    IOError: If ioctl fails for whatever reason.
+    AssertionError: If you're using this code to hack an alien mothership
+                    and their EFI uses a different descriptor size than 0x30
+                    this will blow up.
+    """
+    mmap_size = pmem_get_mmap_size(fd)
+    mmap_desc_size = pmem_get_mmap_desc_size(fd)
+    mmap = ctypes.create_string_buffer(mmap_size * mmap_desc_size)
+
+    err = fcntl.ioctl(fd, PMEM_IOCTL_GET_MMAP, ctypes.pointer(mmap), True)
+    if err:
+        raise IOError("Error (%d) filling mmap buffer." % err)
+
+    assert(mmap_desc_size == 0x30,
+           ("EFI reports descriptor size of 0x%x. This code only knows how "
+            "to handle descriptors 0x30 bytes in length.") % mmap_desc_size)
+
+    return mmap, mmap_size, mmap_desc_size
+
+
+def pmem_parse_mmap(fd):
+    """Retrieve and parse the physical memory map from the Pmem driver.
+
+    Yields: tuples of (start, number of pages, type)
+    """
+    mmap, size, desc_size = pmem_get_mmap(fd)
+    session = rekall_session.Session()
+    buffer_as = addrspace.BufferAddressSpace(data=mmap.raw, session=session)
+    session.SetParameter("default_address_space", buffer_as)
+    profile = EFIProfile(session=session)
+
+    for offset in xrange(0, size, desc_size):
```

Comment:
This does not look right - do you mean to define an array of EFI_MEMORY_DESCRIPTORs?